draft-ietf-quic-transport.md

```diff
@@ -499,43 +499,58 @@ The fields in a Regular packet past the Common Header are the following:
 ~~~
 {: #regular-packet-frames title="Contents of Encrypted Payload"}
 
-### Packet Number Compression and Reconstruction
 
-The complete packet number is a 64-bit unsigned number and is used as part of a
-cryptographic nonce for packet encryption.  To reduce the number of bits
-required to represent the packet number over the wire, at most 48 bits of the
-packet number are transmitted over the wire.  A QUIC endpoint MUST NOT reuse a
-complete packet number within the same connection (that is, under the same
-cryptographic keys).  If the total number of packets transmitted in this
-connection reaches 2^64 - 1, the sender MUST close the connection by sending a
+## Packet Numbers
+
+The packet number is a 64-bit unsigned number and is used as part of a
+cryptographic nonce for packet encryption.  Each endpoint maintains a separate
+packet number for sending and receiving.  The packet number for sending MUST
+increase by one after sending any packet.
+
+A QUIC endpoint MUST NOT reuse a packet number within the same connection (that
+is, under the same cryptographic keys).  If the packet number for sending
+reaches 2^64 - 1, the sender MUST close the connection by sending a
 CONNECTION_CLOSE frame with the error code QUIC_SEQUENCE_NUMBER_LIMIT_REACHED
-(connection termination is described in {{termination}}.)  For unambiguous
-reconstruction of the complete packet number by a receiver from the lower-order
-bits, a QUIC sender MUST NOT have more than 2^(packet_number_size - 2) in flight
-at any point in the connection.  In other words,
+(connection termination is described in {{termination}}.)
+
+To reduce the number of bits required to represent the packet number over the
+wire, only the least significant bits of the packet number are transmitted over
+the wire, up to 48 bits.  The actual packet number for each packet is
+reconstructed at the receiver based on the largest packet number received on a
+successfully authenticated packet.
+
+A packet number is decoded by finding the packet number value that is closest to
+the next expected packet.  The next expected packet is the highest received
```

Comment:
The smallest packet a receiver is waiting for is relatively well defined, but I don't think it's what we want.  The goal of latching to the largest packet is to favor forward progress over handling unlimited reordering.  Otherwise a receiver may never give up on waiting for a packet, and the sender won't know that, causing issues.  Whereas the sender knows the largest packet the receiver has acked.

The sender has all the information to decide when it sends a packet how large the packet number space needs to be.  When coming out of quiescence, it needs to make the packet number slightly larger than otherwise necessary to accommodate large scale reordering, but that's a transitory issue and easy to implement.